src/passes/RemoveUnusedFunctions.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2016 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Removes functions that are never used.
+//
+
+
+#include <memory>
+
+#include "wasm.h"
+#include "pass.h"
+#include "ast_utils.h"
+
+namespace wasm {
+
+struct RemoveUnusedFunctions : public Pass {
+  void run(PassRunner* runner, Module* module) override {
+    std::vector<Function*> root;
+    // Module start is a root.
+    if (module->start.is()) {
+      root.push_back(module->getFunction(module->start));
+    }
+    // Exports are roots.
+    for (auto& curr : module->exports) {
+      root.push_back(module->getFunction(curr->value));
+    }
+    // For now, all functions that can be called indirectly are marked as roots.
+    for (auto& curr : module->table.names) {
+      root.push_back(module->getFunction(curr));
+    }
+    // Compute function reachability starting from the root set.
+    DirectCallGraphAnalyzer analyzer(module, root);
+    // Remove unreachable functions.
+    auto it = module->functions.begin();
```

Comment:
this is `O(N^2)`, if I am not mistaken?

some projects can have tons of tiny functions, so I think we should keep this linear. In e.g. Vacuum.cpp we use a simple skipping pattern,

````
    int skip = 0;
    for (int z = 0; ..  ; z++) {
      if (skip it) {
        skip++;
      } else {
        if (skip > 0) {
          list[z - skip] = list[z];
        }
      }
    .. after loop, resize if skip > 0
````

Maybe we could even make a utility function that does that, and use it here? Or maybe it's too trivial.